gnovm/cmd/gno/depgraph.go

```diff
@@ -0,0 +1,199 @@
+			err := buildGraphData(pkg, allPkgs, make(map[string]bool), make(map[string]bool), &graphData)
+
+			if err != nil {
+				return fmt.Errorf("error in building graph: %w", err)
+			}
+		}
+
+		file, err := os.Create(output)
+		if err != nil {
+			return fmt.Errorf("couldn't open output file: %w", err)
+		}
+		graphFileData := fmt.Sprintf("Digraph G {\nrankdir=\"LR\"\nranksep=20\n%s\n%s\n}", nodeData, graphData)
+		file.Write([]byte(graphFileData))
+		file.Close()
+	} else { // useful for testing - makes a separate graph for each found package
+		if !osm.DirExists(output) {
+			err := os.MkdirAll(output, os.ModePerm)
+			if err != nil {
+				return fmt.Errorf("couldn't make output dir: %w", err)
+			}
+		}
+
+		// decided not to use colors or layouts here since it's a simple graph
+
+		for _, pkg := range allPkgs {
+			pkgPath, err := filepath.Abs(pkg.Dir)
+			if err != nil {
+				return fmt.Errorf("error in getting path of pkg: %w", err)
+			}
+
+			pkgPath = strings.TrimPrefix(pkgPath, rootDir)
+			pkgPath = strings.TrimSuffix(pkgPath, string([]rune{os.PathSeparator}))
+
+			if verbose {
+				fmt.Fprintf(io.Err(), "Generating graph for %q...\n", pkgPath)
+			}
+
+			graphData := ""
+			graphPath := filepath.Join(output, pkgPath)
+			graphPath = graphPath + ".dot"
+			basePath := path.Dir(graphPath)
+			err = os.MkdirAll(basePath, os.ModePerm)
+			if err != nil {
+				return fmt.Errorf("error in making dir for graph: %w", err)
+			}
+
+			file, err := os.Create(graphPath)
+			if err != nil {
+				return fmt.Errorf("couldn't create output file: %w", err)
+			}
+
+			err = buildGraphData(pkg, allPkgs, make(map[string]bool), make(map[string]bool), &graphData)
+
+			if err != nil {
+				return fmt.Errorf("error in building graph: %w", err)
+			}
+
+			graphFileData := fmt.Sprintf("Digraph G {%s}\n", graphData)
+			file.Write([]byte(graphFileData))
+			file.Close()
+		}
+	}
+
+	return nil
+}
+
+// walk through all requires recursively and note dependencies
+func buildGraphData(pkg gnomod.Pkg, allPkgs []gnomod.Pkg, visited map[string]bool, onStack map[string]bool, graphData *string) error {
+	if onStack[pkg.Name] {
+		return fmt.Errorf("cycle detected: %s", pkg.Name)
+	}
+	if visited[pkg.Name] {
+		return nil
+	}
+
+	visited[pkg.Name] = true
+	onStack[pkg.Name] = true
+
+	for _, req := range pkg.Requires {
+		found := false
+
+		for _, candidate := range allPkgs {
+			if candidate.Name != req {
+				continue
+			}
+			if err := buildGraphData(candidate, allPkgs, visited, onStack, graphData); err != nil {
+				return err
+			}
+			found = true
+			// this check is wildly inefficient - change this to not work with strings directly
+			if !strings.Contains(*graphData, "\""+pkg.Name+"\" -> \""+req+"\"\n") {
+				*graphData += "\"" + pkg.Name + "\" -> \"" + req + "\"\n"
+			}
+		}
+		if !found {
+			return fmt.Errorf("couldn't find dependency %q for package %q", req, pkg.Name)
+		}
+	}
+
+	onStack[pkg.Name] = false
+
+	return nil
+}
```

Comment:
You don't need to check if the line already exists because you are already ensuring that you visit the package and requirements only once.
 
```suggestion
func buildGraphData(pkg gnomod.Pkg, allPkgs []gnomod.Pkg, visited map[string]bool, onStack map[string]bool, graphData *string) error {
	if onStack[pkg.Name] {
		return fmt.Errorf("cycle detected: %s", pkg.Name)
	}
	if visited[pkg.Name] {
		return nil
	}

	visited[pkg.Name] = true
	onStack[pkg.Name] = true

	for _, req := range pkg.Requires {
		var candidate gnomod.Pkg
		var found bool
		for _, candidate = range allPkgs {
			if candidate.Name != req {
				continue
			}

			found = true
			break
		}
		if !found {
			return fmt.Errorf("couldn't find dependency %q for package %q", req, pkg.Name)
		}

		if err := buildGraphData(candidate, allPkgs, visited, onStack, graphData); err != nil {
			return err
		}

		*graphData += "\"" + pkg.Name + "\" -> \"" + req + "\"\n"
	}

	onStack[pkg.Name] = false

	return nil
}
```